### PR TITLE
v2 Java changes for consistency with other SDKs

### PR DESCRIPTION
**Pull request by jkodumal**
I'll walk through the changes in comments below.

### CODE REVIEW

File: src/main/java/com/launchdarkly/client/EvaluationError.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.launchdarkly.client;
+
+/**
+ * An error indicating an abnormal result from evaluating a feature
+ */
+public class EvaluationError extends Exception {
```
Comment (jkodumal) [inline src/main/java/com/launchdarkly/client/EvaluationError.java]: A new exception to help us distinguish precondition / well-formedness failures during evaluation from `null` evaluation results, which should result in the `offVariation` being served.

File: src/main/java/com/launchdarkly/client/FeatureFlag.java [hunk 2]
```diff
@@ -61,33 +61,32 @@ JsonElement getOffVariationValue() {
     return null;
   }
 
-  EvalResult evaluate(LDUser user, FeatureStore featureStore) {
+  EvalResult evaluate(LDUser user, FeatureStore featureStore) throws EvaluationError {
     if (user == null || user.getKey() == null) {
```
Comment (jkodumal) [inline src/main/java/com/launchdarkly/client/FeatureFlag.java]: In this case we should return the default.

File: src/main/java/com/launchdarkly/client/FeatureFlag.java [hunk 3]
```diff
@@ -61,33 +61,32 @@ JsonElement getOffVariationValue() {
...
       FeatureFlag prereqFeatureFlag = featureStore.get(prereq.getKey());
       JsonElement prereqEvalResult = null;
       if (prereqFeatureFlag == null) {
         logger.error("Could not retrieve prerequisite flag: " + prereq.getKey() + " when evaluating: " + key);
         return null;
       } else if (prereqFeatureFlag.isOn()) {
-        prereqEvalResult = prereqFeatureFlag.evaluate(user, featureStore, events, visited);
-        if (prereqEvalResult == null || !prereqEvalResult.equals(prereqFeatureFlag.getVariation(prereq.getVariation()))) {
+        prereqEvalResult = prereqFeatureFlag.evaluate(user, featureStore, events);
+        try {
```
Comment (jkodumal) [inline src/main/java/com/launchdarkly/client/FeatureFlag.java]: If prerequisite evaluation fails for whatever reason, we return the off variation, not an error. So we don't bubble up prereq evaluation exceptions.

File: src/main/java/com/launchdarkly/client/FeatureFlag.java [hunk 4]
```diff
@@ -123,10 +122,18 @@ private Integer evaluateIndex(LDUser user) {
     return fallthrough.variationIndexForUser(user, key, salt);
   }
 
-  private JsonElement getVariation(Integer index) {
-    if (index == null || index >= variations.size()) {
+  private JsonElement getVariation(Integer index) throws EvaluationError {
```
Comment (jkodumal) [inline src/main/java/com/launchdarkly/client/FeatureFlag.java]: We distinguish the fallthrough rule not matching (index is null) from the index not referencing a valid variation, which would indicate that the rules aren't well-formed.

File: src/main/java/com/launchdarkly/client/LDClient.java [hunk 5]
```diff
@@ -294,47 +294,36 @@ public JsonElement jsonVariation(String featureKey, LDUser user, JsonElement def
...
       FeatureFlag featureFlag = config.featureStore.get(featureKey);
-      if (featureFlag != null) {
-        if (config.stream && config.debugStreaming) {
-          FeatureFlag pollingResult = requestor.makeRequest(featureKey, true);
-          if (!featureFlag.equals(pollingResult)) {
-            logger.warn("Mismatch between streaming and polling feature! Streaming: {} Polling: {}", featureFlag, pollingResult);
-          }
-        }
-      } else {
+      if (featureFlag == null) {
```
Comment (jkodumal) [inline src/main/java/com/launchdarkly/client/LDClient.java]: I removed the `debugStreaming` option; I believe it has outlived its usefulness.

File: src/main/java/com/launchdarkly/client/LDClient.java [hunk 6]
```diff
@@ -294,47 +294,36 @@ public JsonElement jsonVariation(String featureKey, LDUser user, JsonElement def
...
             return evalResult.getValue();
           }
-        }
-      } else {
-        JsonElement offVariation = featureFlag.getOffVariationValue();
-        if (offVariation != null) {
-          return offVariation;
-        }
+      }
+      JsonElement offVariation = featureFlag.getOffVariationValue();
```
Comment (jkodumal) [inline src/main/java/com/launchdarkly/client/LDClient.java]: Return the `offVariation` unless there is an error.

File: src/main/java/com/launchdarkly/client/Operator.java [hunk 7]
```diff
@@ -25,13 +25,6 @@ public boolean apply(JsonPrimitive uValue, JsonPrimitive cValue) {
       if (uValue.isNumber() && cValue.isNumber()) {
         return uValue.getAsDouble() == cValue.getAsDouble();
       }
```
Comment (jkodumal) [inline src/main/java/com/launchdarkly/client/Operator.java]: This cleans up the semantics of the `in` operator.

File: src/main/java/com/launchdarkly/client/FeatureFlag.java [hunk 8]
```diff
@@ -61,33 +61,32 @@ JsonElement getOffVariationValue() {
...
+    JsonElement value = evaluate(user, featureStore, prereqEvents);
     return new EvalResult(value, prereqEvents);
   }
 
   // Returning either a JsonElement or null indicating prereq failure/error.
-  private JsonElement evaluate(LDUser user, FeatureStore featureStore, List<FeatureRequestEvent> events, Set<String> visited) {
+  private JsonElement evaluate(LDUser user, FeatureStore featureStore, List<FeatureRequestEvent> events) throws EvaluationError {
     boolean prereqOk = true;
-    visited.add(key);
     for (Prerequisite prereq : prerequisites) {
```
Comment (jkodumal) [inline src/main/java/com/launchdarkly/client/FeatureFlag.java]: Removed the broken cycle detection algorithm.

File: src/main/java/com/launchdarkly/client/EvaluationError.java (hunk 1 above)
Comment (drichelson) [inline src/main/java/com/launchdarkly/client/EvaluationError.java]: This should not be public- we don't throw it from any public method. 
Changing the name to EvaluationException is more consistent with Java naming conventions, unless we ARE going to throw this from a public method (probly not  a good idea), then we should name it LDEvaluationException or similar.

File: src/main/java/com/launchdarkly/client/FeatureFlag.java [hunk 9]
```diff
@@ -61,33 +61,32 @@ JsonElement getOffVariationValue() {
...
-        prereqEvalResult = prereqFeatureFlag.evaluate(user, featureStore, events, visited);
-        if (prereqEvalResult == null || !prereqEvalResult.equals(prereqFeatureFlag.getVariation(prereq.getVariation()))) {
+        prereqEvalResult = prereqFeatureFlag.evaluate(user, featureStore, events);
+        try {
+          JsonElement variation = prereqFeatureFlag.getVariation(prereq.getVariation());
+          if (prereqEvalResult == null || variation == null || !prereqEvalResult.equals(variation)) {
+            prereqOk = false;
+          }
+        } catch (EvaluationError err) {
           prereqOk = false;
```
Comment (drichelson) [inline src/main/java/com/launchdarkly/client/FeatureFlag.java]: I think we should log the exception here- at debug or warn.

File: src/main/java/com/launchdarkly/client/LDClient.java (hunk 5 above)
Comment (drichelson) [inline src/main/java/com/launchdarkly/client/LDClient.java]: 🙏

File: src/main/java/com/launchdarkly/client/FeatureFlag.java [hunk 10]
```diff
@@ -61,33 +61,32 @@ JsonElement getOffVariationValue() {
     return null;
   }
 
-  EvalResult evaluate(LDUser user, FeatureStore featureStore) {
+  EvalResult evaluate(LDUser user, FeatureStore featureStore) throws EvaluationError {
```
Comment (drichelson) [inline src/main/java/com/launchdarkly/client/FeatureFlag.java]: I realized the changes you made were to solve a specific problem, but If we're moving to this pattern of throwing exceptions during evaluation we should consider change it across the board for consistency. One example is Line 57 in this file: getOffVariationValue().

Comment (jkodumal) [inline src/main/java/com/launchdarkly/client/FeatureFlag.java]: Good catch.

File: src/main/java/com/launchdarkly/client/FeatureFlag.java (hunk 9 above)
Comment (jkodumal) [inline src/main/java/com/launchdarkly/client/FeatureFlag.java]: Good catch.

**Comment by jkodumal** [comment]
@drichelson think I've covered all the PR feedback. Can you circle back when you have a chance?

File: src/main/java/com/launchdarkly/client/EvaluationError.java [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package com.launchdarkly.client;
+
+/**
+ * An error indicating an abnormal result from evaluating a feature
+ */
+class EvaluationError extends Exception {
```
Comment (drichelson) [inline src/main/java/com/launchdarkly/client/EvaluationError.java]: My original comment was 2 parts, the 2nd part was not addressed/discussed:
Changing the name to EvaluationException is more consistent with Java naming conventions.

**Comment by jkodumal** [comment]
Should be all set now.

**Comment by drichelson** [comment]
👍

**Pull request merged**